test/jdk/java/net/httpclient/HttpResponseLimitingTest.java

```diff
@@ -0,0 +1,396 @@
+                        try (var outputStream = exchange.getResponseBody()) {
+                            outputStream.write(RESPONSE_BODY);
+                        }
+                        exchange.close();
+                    },
+                    handlerPath);
+
+            // Create the client and the request
+            HttpClient client = createClient(version, sslContext);
+            HttpRequest request = HttpRequest.newBuilder(requestUri).version(version).build();
+
+            // Create the pair
+            return new ServerClientPair(server, client, request);
+
+        }
+
+        private static HttpTestServer createServer(HttpClient.Version version, SSLContext sslContext) {
+            try {
+                return HttpTestServer.create(version, sslContext);
+            } catch (IOException exception) {
+                throw new UncheckedIOException(exception);
+            }
+        }
+
+        private static HttpClient createClient(HttpClient.Version version, SSLContext sslContext) {
+            HttpClient.Builder builder = HttpClient.newBuilder().version(version).proxy(NO_PROXY);
+            if (sslContext != null) {
+                builder.sslContext(sslContext);
+            }
+            return builder.build();
+        }
+
+        private <T> HttpResponse<T> request(BodyHandler<T> downstreamHandler, long capacity) throws Exception {
+            var handler = BodyHandlers.limiting(downstreamHandler, capacity);
+            return client.send(request, handler);
+        }
+
+        @Override
+        public String toString() {
+            String version = client.version().toString();
+            return client.sslContext() != null ? version.replaceFirst("_", "S_") : version;
+        }
+
+    }
+
+    @AfterAll
+    static void closeServerClientPairs() {
+        for (var pair : new ServerClientPair[]{HTTP1, HTTPS1, HTTP2, HTTPS2}) {
+            pair.client.close();
+            pair.server.stop();
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("sufficientCapacities")
+    void testSuccessOnSufficientCapacityForByteArray(ServerClientPair pair, long sufficientCapacity) throws Exception {
+        HttpResponse<byte[]> response = pair.request(BodyHandlers.ofByteArray(), sufficientCapacity);
+        verifyHeaders(response.headers());
+        assertArrayEquals(RESPONSE_BODY, response.body());
+    }
+
+    @ParameterizedTest
+    @MethodSource("sufficientCapacities")
+    void testSuccessOnSufficientCapacityForInputStream(ServerClientPair pair, long sufficientCapacity) throws Exception {
+        HttpResponse<InputStream> response = pair.request(BodyHandlers.ofInputStream(), sufficientCapacity);
+        verifyHeaders(response.headers());
+        try (InputStream responseBodyStream = response.body()) {
+            byte[] responseBodyBuffer = new byte[RESPONSE_BODY.length];
+            int responseBodyBufferLength = responseBodyStream.read(responseBodyBuffer);
```

Comment:
I think a better way would be to read the `responseBodyStream` till EOF and keep accumulating the read data into the byte array. After reaching EOF, we then compare the accumulated byte array for the expected size and content. In the current form, if there's a short read (i.e. this `read()` call reads lesser bytes than the `responseBodyBuffer.length`, even if there's additional data available), then this test will fail.

Reply:
Implemented in 21ae1079aaa3730317ef1bb11277e929dbd6174d.